codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,457 @@
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /**
+     * Construct a new handler whose child channels run in a different event loop.
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
+     * @param streamGroup event loop for registering child channels
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
+    }
+
+    Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup, Http2FrameWriter frameWriter) {
+        if (!streamHandler.getClass().isAnnotationPresent(Sharable.class)) {
+            throw new IllegalArgumentException("streamHandler must be Sharable");
+        }
+        this.streamHandler = streamHandler;
+        this.streamGroup = streamGroup;
+        http2Handler = new InternalHttp2ConnectionHandler(new DefaultHttp2Connection(server),
+                new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER),
+                new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER),
+                new FrameListener());
+        http2Handler.connection().addListener(new ConnectionListener());
+        streamInfoKey = http2Handler.connection().newKey();
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
+    }
+
+    /**
+     * Save context and load any dependencies.
+     */
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        String name = ctx.name() + "-Http2ConnectionHandler";
+        if (ctx.pipeline().context(name) != null) {
+          throw new Exception("Name conflict: " + name);
+        }
+        ctx.pipeline().addBefore(ctx.invoker(), ctx.name(), name, http2Handler);
+        this.http2HandlerCtx = ctx.pipeline().context(http2Handler);
+    }
+
+    /**
+     * Clean up any dependencies.
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+      ctx.pipeline().remove(http2Handler);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+            return;
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        ctx.fireUserEventTriggered(upgrade.retain());
```

Comment:
Separating out the `retain()` is misleading. The `release()` at the end is not related to the `retain()`. Instead, it is to make sure that we release the reference handed to us. If we are concerned about `fireUserEventTriggered` throwing, then it would look more like:

```java
UpgradeEvent upgrade = (UpgradeEvent) evt;
try {
    ctx.fireUserEventTriggered(upgrade.retain());
    ...the rest of the method...
} finally {
    upgrade.release();
}
```

I've been treating all the `fire\*` methods as non-throwing, since the call will `fireExceptionCaught` instead. It seems `DefaultChannelHandlerInvoker` isn't concerned about OutOfMemory errors, such that if an exception occurs I wouldn't know whether I was supposed to release the reference or not. Is treating it non-throwing improper? If so, is it ever safe to try to keep going? For example, sometimes I will loop over all streams and call a `fire\*` method for each; how would that be handled?